pkg/validatorfactory/schema_patcher.go

```diff
@@ -0,0 +1,141 @@
+package validatorfactory
+
+import (
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+	"sigs.k8s.io/kubectl-validate/pkg/utils"
+)
+
+type SchemaPatch struct {
+	Slug        string
+	Description string
+
+	// (Inclusive) version range for which this patch applies
+	MinMinorVersion int
+	MaxMinorVersion int
+
+	// Nil is wildcard
+	AppliesToGV func(schema.GroupVersion) bool
+	Transformer utils.SchemaVisitor
+}
+
+var zero int64 = int64(0)
+var schemaPatches []SchemaPatch = []SchemaPatch{
+	{
+		Slug:            "AllowEmptyByteFormat",
+		Description:     "Work around discrepency between treatment of native vs CRD `byte` strings. Native types allow empty, CRDs do not",
+		MinMinorVersion: 0,
+		MaxMinorVersion: 0,
+		AppliesToGV:     nil,
+		Transformer: utils.PostorderVisitor(func(ctx utils.VisitingContext, s *spec.Schema) bool {
+			if s.Format != "byte" || len(s.Type) != 1 || s.Type[0] != "string" {
+				return true
+			}
+
+			// Change format to "", and add new `$and: {$or: [{format: "byte"}, {maxLength: 0}]}
+			s.AllOf = append(s.AllOf, spec.Schema{
+				SchemaProps: spec.SchemaProps{
+					AnyOf: []spec.Schema{
+						{
+							SchemaProps: spec.SchemaProps{
+								Format: s.Format,
+							},
+						},
+						{
+							SchemaProps: spec.SchemaProps{
+								MaxLength: &zero,
+							},
+						},
+					},
+				},
+			})
+			s.Format = ""
+			return true
+		}),
+	},
+	{
+		Slug:        "AllowEmptyDateTime",
+		Description: "metav1.Time published OpenAPI definitions do not allow empty/null, but Kubernetes in practice does.",
+		Transformer: utils.PostorderVisitor(func(ctx utils.VisitingContext, s *spec.Schema) bool {
+			if s.Format != "date-time" || len(s.Type) != 1 || s.Type[0] != "string" {
+				return true
+			}
+
+			// Change format to "", and add new `$and: {$or: [{format: "date-time"}, {maxLength: 0}]}
+			s.AllOf = append(s.AllOf, spec.Schema{
```

Comment:
Could we refactor this transformation to be idempotent? i.e. if you run it multiple times against the same schema it will modify it each time. Would be better if it only performed the transformation once.

Reply:
Fixed